Docs for sandbox usage (#1477)
AztecProtocol/aztec-packages: docs/docs/dev_docs/sandbox/main.md

title: Aztec Sandbox
---

import Image from "@theme/IdealImage";
import GithubCode from '../../../src/components/GithubCode';

## Introduction

> The Aztec Sandbox aims to provide a complete layer 2 system against which DApp developers can build and deploy contracts in a fast, safe and free environment.

Reviewer: ```suggestion
The Aztec Sandbox aims to provide a local development system against which you can build and test Noir contracts in a fast, safe, and free environment.
```
Though the _fast_ part is still pending :-P

## What do you need?

- Node.Js >= v18
- Docker and Docker Compose (Docker Desktop under WSL2 on windows)

That's it...

## Ok, so how do I try it out?

Well, you can find instructions [at the website](https://up.aztec.network).

Or you can just curl the site instead like this:

`/bin/bash -c "$(curl -fsSL https://up.aztec.network)"`

It will download and execute a script invoking docker compose with 2 containers:

- Anvil
- Aztec Sandbox

It will need to create servers on localhost ports 8545 (Anvil) and 8080 (Sandbox) so you will need to ensure nothing conflicts with this.

Within a few seconds the Sandbox should be ready for use!

<Image img={require("/img/sandbox.png")} />

## Great, but I want to know more about it

Aztec's Layer 2 network is a fully programmable combined private/public ZK rollup. To achieve this, the network contains the following primary components:

- Archiver - Syncs with L1 and maintains a read-only repository of on-L1 state.
- World State - The collection of Merkle Trees.
- P2P - Currently just a locally stored collection of pending transactions.

> - Sequencer - Responsible for ordering transactions into the rollup, executing the rollup circuits and publishing blocks.

Reviewer: I'd remove these from the quickstart, we can expand more on these in an architecture section. A dev here is probably just concerned with node/rpc/client.

Author: Yeah I did wonder about that myself.

- Aztec.js - Aztec's client library for interacting with the Aztec RPC Server (think Ethers.js).

All of this is included in the Sandbox, with the exception of Aztec.js which you can use to interact with it.

With the help of Aztec.js you will be able to:

- Create an account
- Deploy a contract
- Call view methods on contracts
- Simulate the calling of contract functions
- Send transactions to the network
- Be notified when transactions settle
- Retrieve and view unencrypted logs emitted by contracts
- Query chain state such as chain id, block number etc.

## I have the Sandbox running, show me how to use it!

Let's have a complete walkthrough from start to finish. I'm using WSL2 Ubuntu under Windows but the following should work under regular Linux or MacOS. We will deploy and use a private token contract on our Sandbox. Writing the contract itself is out of scope for this tutorial, we will use the Private Token Contract supplied as one of the example contracts.

Let's create an empty project called `private-token`. If you are familiar with setting up Typescript projects then you can fast-forward the next couple of steps.

```

> phil@LAPTOP-E72241SF:~$ node -v

Reviewer: ```suggestion
$ node -v
```
Just to clean this up

v18.8.0
phil@LAPTOP-E72241SF:~$ mkdir private-token
phil@LAPTOP-E72241SF:~$ cd private-token

> phil@LAPTOP-E72241SF:~/private-token$ yarn init

Reviewer: `yarn init -y` is your friend for this kind of examples

Reviewer: Also, should we clarify that we'll be using `yarn` for this example, but devs can also use `npm` if they want to?

```
{
  "compilerOptions": {
    "outDir": "dest",
    "rootDir": "src",
    "target": "es2020",
    "lib": ["dom", "esnext", "es2017.object"],
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "strict": true,
    "declaration": true,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "downlevelIteration": true,
    "inlineSourceMap": true,
    "declarationMap": true,
    "importHelpers": true,
    "resolveJsonModule": true,
    "composite": true,
    "skipLibCheck": true
  },
  "references": [],
  "include": ["src"]
}
```

Add a `scripts` section to `package.json` and set `"type": "module"`:

```
{
	"name": "private-token",
	"version": "1.0.0",
	"description": "My first private token contract",
	"main": "index.js",
	"author": "Phil",
	"license": "MIT",
	"type": "module",
	"scripts": {
		"build": "yarn clean && tsc -b",
		"build:dev": "tsc -b --watch",
		"clean": "rm -rf ./dest tsconfig.tsbuildinfo",
		"start": "yarn build && export DEBUG='private-token' && node ./dest/index.js"
	},
	"devDependencies": {
		"@types/node": "^20.4.9",
		"typescript": "^5.1.6"
	}
}
```

Now we want to install 2 Aztec packages from npm:

```

> yarn add @aztec/noir-contracts
> yarn add @aztec/aztec.js

Reviewer: ```suggestion
yarn add @aztec/aztec.js @aztec/noir-contracts
```

Running `yarn start` should give:

```
  private-token Aztec Sandbox Info  { version: 1, chainId: 31337 } +0ms
```

Great!. The Sandbox is running and we are able to interact with it.

## Account Creation/Deployment

The next step is to create some accounts. I won't go into detail about accounts as that is covered [here](../../concepts/foundation/accounts/main.md). But creating an account on the Sandbox does 2 things:

1. Deploys an account contract reprepresenting you allowing you to perform actions on the network (deploy contracts, call functions etc).
2. Adds your encryption keys to the RPC Server allowing it to decrypt and manage your private state.

Continue with adding the following to our example:

#include_code Accounts /docs/src/code_examples/sandbox_example.ts typescript

Running `yarn start` should now output:

```
  private-token Aztec Sandbox Info  { version: 1, chainId: 31337 } +0ms
  private-token Creating accounts using schnorr signers... +2ms
  private-token Created Alice's account at 0x054d89d0...f17e +23s
  private-token Created Bob's account at 0x0a8410a1...7c48 +1ms
```

That might seem like a lot to digest but it can be broken down into the following steps:

> 1. We create 2 `Account` objects in Typescript. This object heavily abstracts away the mechanics of configuring and deploying an account contract and setting up a 'wallet' for signing transactions. If you aren't interest in building new types of account contracts or wallets then you don't need to be too concerned with it. In this example we have constructed account contracts and corresposing wallets that sign/verify transactions using schnorr signatures.

Reviewer: ```suggestion
1. We create 2 `Account` objects in Typescript. This object heavily abstracts away the mechanics of configuring and deploying an account contract and setting up a 'wallet' for signing transactions. If you aren't interested in building new types of account contracts or wallets then you don't need to be too concerned with it. In this example we have constructed account contracts and corresposing wallets that sign/verify transactions using schnorr signatures.
```

  private-token Transaction status is mined +8s
  private-token Contract successfully deployed at address 0x143e0af4...11b6 +7ms
  private-token Retrieving unencrypted logs for block 3 +4ms
  private-token Emitted logs:  [ 'Balance set in constructor' ] +5ms
  private-token Alice's balance 1000000 +4s
  private-token Bob's balance 0 +3s
  private-token Transferring 543 tokens from Alice to Bob... +0ms
  private-token Retrieving unencrypted logs for block 4 +20s
  private-token Emitted logs:  [ 'Coins transferred' ] +13ms
  private-token Alice's balance 999457 +4s
  private-token Bob's balance 543 +3s
  private-token Minting 10000 tokens to Bob... +1ms
  private-token Retrieving unencrypted logs for block 5 +13s
  private-token Emitted logs:  [ 'Coins minted' ] +13ms
  private-token Alice's balance 999457 +4s
  private-token Bob's balance 10543 +4s
```

That's it! We have successfully deployed a private token contract to an instance of the Aztec network and mined private state-transitioning transactions. We have also queried the resulting state all via the interfaces provided by the contract.

> ## Accounts and Keys

Author: I kind of felt the same about that section. The reason I used 2 accounts is that without it you can't see the results of the transfer, i.e. Bob's balance.

It's a compromise really. In a private transaction you can only see one side of it but that doesn't make a very good demo!

> ## Accounts and Keys

Reviewer: > but that doesn't make a very good demo

I disagree - I think it's a great demo for a private-first chain! Otherwise this demo is exactly the same as a regular public ERC20.
